content/blog/graphql-as-a-gateway/index.md

```diff
@@ -0,0 +1,217 @@
+
+## 🙋 Why GraphQL and not REST?
+
+GraphQL was initially developed by Facebook, [been open-sourced in 2015](https://code.fb.com/core-data/graphql-a-data-query-language/). Many companies started using it for internal APIs, but some are already exposing it as its public API (eg: GitHub, Shopify, Yelp and Contentful).
+
+### Schema validation and documentation
+
+In GraphQL, differently from REST, a schema is always required, following [GraphQL Foundation](https://graphql.org) directives. This allows not only schema validation since day one, but documentation as well. Typed languages can benefit from the schema and generate type definitions through it (eg: [typescript graphql-code-generator](https://github.com/dotansimha/graphql-code-generator)).
+
+```graphql
+type Query {
+  """
+  Fetch posts by user, allowing custom filtering
+  """
+  posts(input: PostQueryInput): [Post]
+}
+
+input PostQueryInput {
+  categoryId: String
+}
+
+"""
+Post bla bla bla
+"""
+type Post {
+  title: String
+  categories: [Category]
+}
+
+type Category {
+  name: String
+}
+```
+
+In theory, the same could be done in REST by using OpenAPI/Swagger, specially for documentation ~~but yaml, yikes~~. But, the code generation tools are not exactly the best and sometimes the schema validation does not work properly, requiring extra middlewares and tweaks around it. As GraphQL is standardised in this sense, it doesn't suffer from these issues.
+
+### An endpoint to rule them all
+
+GraphQL only exposes one endpoint. Instead of having N client requests, only one is needed, with the server orchestrating everything required to fulfil the whole request, such as calling multiple micro-services -- in any protocol, mapping outputs to the expected schema and logging requests/responses. Besides, this solves [under-fetching issues](https://stackoverflow.com/questions/44564905/what-is-over-fetching-or-under-fetching/44568365).
+
+### Out-of-box standards
+
+REST has been around for a long time, and during this period developers started to have specific necessities, such as sparse fieldsets, versioning, pagination. There are a lot of ways of doing these, but none is really a standard (the closest solution is [JSONAPI](https://jsonapi.org/)). GraphQL comes with some of these specs out-of-box:
+
+- [Sparse fieldset](https://graphql.org/learn/queries/#fields): while making the request, all the required fields need to be specified, avoiding over-fetching
+- [Versioning](https://graphql.org/learn/best-practices/#versioning): versioning is not a good practice in GraphQL instead, the schema should continually evolute. New capabilities can be inserted on new types and fields, allowing the client to plan and decide when to change to new resolvers. If a field needs to be deprecated, one can use a directive such as [`@deprecated`](https://www.apollographql.com/docs/graphql-tools/schema-directives).
```

Comment:
You seem to say that the versioning is good but you also say that it's not a good practice. Which one is the right one?

Reply:
Route versioning is bad in GraphQL... Added "route versioning" in the beginning of it. Versioning kinda happen, but using new resolvers/types instead